crates/ide-assists/src/handlers/explicit_enum_discriminant.rs

```diff
@@ -0,0 +1,202 @@
+use hir::Semantics;
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    source_change::SourceChangeBuilder,
+    RootDatabase,
+};
+use syntax::{ast, AstNode};
+
+use crate::{AssistContext, Assists};
+
+// Assist: explicit_enum_discriminant
+//
+// Adds explicit discriminant to all enum variants.
+//
+// ```
+// enum TheEnum$0 {
+//     Foo,
+//     Bar,
+//     Baz = 42,
+//     Quux,
+// }
+// ```
+// ->
+// ```
+// enum TheEnum {
+//     Foo = 0,
+//     Bar = 1,
+//     Baz = 42,
+//     Quux = 43,
+// }
+// ```
+pub(crate) fn explicit_enum_discriminant(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let enum_node = ctx.find_node_at_offset::<ast::Enum>()?;
+    let enum_def = ctx.sema.to_def(&enum_node)?;
+
+    let is_data_carrying = enum_def.is_data_carrying(ctx.db());
+    let has_primitive_repr = enum_def.repr(ctx.db()).and_then(|repr| repr.int).is_some();
+
+    // Data carrying enums without a primitive repr have no stable discriminants.
+    if is_data_carrying && !has_primitive_repr {
+        return None;
+    }
+
+    let variant_list = enum_node.variant_list()?;
+
+    // Don't offer the assist if the enum has no variants or if all variants already have an
+    // explicit discriminant.
+    if variant_list.variants().all(|variant_node| variant_node.expr().is_some()) {
+        return None;
+    }
+
+    acc.add(
+        AssistId("explicit_enum_discriminant", AssistKind::RefactorRewrite),
+        "Add explicit enum discriminants",
+        enum_node.syntax().text_range(),
+        |builder| {
+            for variant_node in variant_list.variants() {
+                add_variant_discriminant(&ctx.sema, builder, &variant_node);
+            }
+        },
+    );
+
+    Some(())
+}
+
+fn add_variant_discriminant(
+    sema: &Semantics<'_, RootDatabase>,
+    builder: &mut SourceChangeBuilder,
+    variant_node: &ast::Variant,
+) {
+    if variant_node.expr().is_some() {
+        return;
+    }
+
+    let Some(variant_def) = sema.to_def(variant_node) else {
```

Comment:
The enum discriminant inlay hint does some extra work here, but I'm not sure why that's needed. See [discriminant.rs](https://github.com/rust-lang/rust-analyzer/blob/f454ea877180f0fd8d53390a1d9e6bdd6047b35d/crates/ide/src/inlay_hints/discriminant.rs#L55).

Reply:
`descend_node_into_attributes` is needed to handle macro expansions (especially proc-macro attribute expansions), since if semantic information is acquired for the wrong syntax node it'll cause a panic.

Reply:
I haven't been able to find a case where this makes a difference - putting an attribute macro that does nothing:
```rust
#[proc_macro_attribute]
pub fn identity(_attr: TokenStream, input: TokenStream) -> TokenStream {
    input
}
```
...on an enum disables (breaks?) both this assist and the discriminant inlay hint. The enum variant hover info still works, and gets the discriminant correct even with a macro that inserts a new variant at the start.

Reply:
Odd, that shouldn't prevent inlays. Either way this won't panic as the syntax node is cached in semantics already. It's fine to skip that as it makes things more complicated for now